tidy3d/plugins/near2far/near2far.py

```diff
@@ -1,130 +1,483 @@
+        sim_data: SimulationData,
+        surface: Near2FarSurface,
+        frequency: float,
+        n: float,
+        pts_per_wavelength: int = 10) -> xr.Dataset:
+        """Returns resampled surface current densities associated with the surface monitor.
+
+        Parameters
+        ----------
+        sim_data : :class:`.SimulationData`
+            Container for simulation data containing the near field monitors.
+        surface: :class:`.Near2FarSurface`
+            :class:`.Near2FarSurface` to use as source of near field.
+        frequency : float
+            Frequency to select from each :class:`.FieldMonitor` to use for projection.
+            Must be a frequency stored in each :class:`FieldMonitor`.
+        n : float
+            Real part of the refractive index associated with the background medium.
+        pts_per_wavelength : int
+            Number of points per wavelength with which to discretize the
+            surface monitors for the projection. Default: 10.
+
+        Returns
+        -------
+        xarray.Dataset
+            Colocated surface current densities for the given surface.
         """
 
         try:
-            # fill nans with 0, not sure where nans come from..
-            field_data = sim_data.at_centers(mon_name).fillna(0)
+            field_data = sim_data[surface.monitor.name]
         except Exception as e:
             raise SetupError(
-                f"No data for monitor named '{mon_name}' " "found in supplied sim_data."
+                f"No data for monitor named '{surface.monitor.name}' " "found in sim_data."
             ) from e
 
-        monitor_fields = list(field_data.keys())
-        if any(field_name not in monitor_fields for field_name in ("Ex", "Ey", "Hx", "Hy", "Hz")):
-            raise SetupError(f"Monitor named '{mon_name}' doesn't store all field values")
+        currents = Near2Far._fields_to_currents(field_data, surface)
+        currents = Near2Far._resample_surface_currents(
+            currents, sim_data, surface, frequency, n, pts_per_wavelength
+            )
+
+        return currents
+
+    @staticmethod
+    def _fields_to_currents(
+        field_data: FieldData,
+        surface: Near2FarSurface) -> FieldData:
+        """Returns surface current densities associated with a given :class:`.FieldData` object.
+
+        Parameters
+        ----------
+        field_data : :class:`.FieldData`
+            Container for field data associated with the given near field surface.
+        surface: :class:`.Near2FarSurface`
+            :class:`.Near2FarSurface` to use as source of near field.
+
+        Returns
+        -------
+        :class:`.FieldData`
+            Surface current densities for the given surface.
+        """
+
+        # figure out which field components are tangential or normal to the monitor
+        normal_field, tangent_fields = surface.monitor.pop_axis(('x', 'y', 'z'), axis=surface.axis)
+
+        signs = np.array([-1, 1]) 
```

Comment:
verified that this works?  I was guessing but yea we should definitely retest everything with this implementation to make sure it's all good in general after the changes.